### PR TITLE
LIV-958: Added live captioning reach objects

### PR DESCRIPTION
**Pull request by nadav-ravid**
PR is mostly done. 
Still testing to make sure it's working like i want it

### CODE REVIEW

File: plugins/reach/ReachPlugin.php [hunk 1]
```diff
@@ -368,9 +368,9 @@ public static function shouldSkipRulesValidation($entryId, $ks)
 
 	public static function isEntryTypeSupportedForReach($entryType)
 	{
-    	$supportedEntryTypes = kConf::get('reach_supported_entry_types', 'runtime_config', array(entryType::MEDIA_CLIP));
+		$supportedEntryTypes = kConf::get('reach_supported_entry_types', 'runtime_config', array(entryType::MEDIA_CLIP, entryType::LIVE_STREAM));
```
[the remark this answers is not in the document]
Comment (david-winder-kaltura) [reply plugins/reach/ReachPlugin.php]: Done

File: plugins/reach/lib/api/data/entryVendorTaks/KalturaScheduledVendorTaskData.php [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+
+	public function validateForInsert($propertiesToSkip = array())
+	{
+		$this->validatePropertyNotNull('scheduledEventId');
+		$this->validateScheduledEvent();
+
+		return parent::validateForInsert($propertiesToSkip);
+	}
+
+	private function validateScheduledEvent()
```
[the remark this answers is not in the document]
Comment (david-winder-kaltura) [reply plugins/reach/lib/api/data/entryVendorTaks/KalturaScheduledVendorTaskData.php]: Done

File: plugins/schedule/base/lib/model/LiveStreamScheduleEvent.php [hunk 3]
```diff
@@ -284,4 +284,40 @@ public function getEventTransitionTimes()
...
+	 * Removes feature from event with a given name
+	 *
+	 * Name of feature to remove
+	 * @param string $featureName
+	 *
+	 * @throws PropelException
+	 */
+	public function removeFeature($featureName)
+	{
+		$featureList = $this->getLiveFeatures() ?: [];
```
[the remark this answers is not in the document]
Comment (david-winder-kaltura) [reply plugins/schedule/base/lib/model/LiveStreamScheduleEvent.php]: Done

File: plugins/reach/lib/api/data/entryVendorTaks/KalturaScheduledVendorTaskData.php [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+ 	 */
+	public function toInsertableObject($object_to_fill = null, $props_to_skip = array())
+	{
+		if (is_null($object_to_fill))
+		{
+			$object_to_fill = new kScheduledVendorTaskData();
+		}
+
+		$connectedEvent = BaseScheduleEventPeer::retrieveByPK($this->scheduledEventId);
+
```
[the remark this answers is not in the document]
Comment (david-winder-kaltura) [reply plugins/reach/lib/api/data/entryVendorTaks/KalturaScheduledVendorTaskData.php:55]: This is checked during the validate for insert

File: plugins/reach/lib/api/data/entryVendorTaks/KalturaScheduledVendorTaskData.php [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+		$vendorCatalogItem = VendorCatalogItemPeer::retrieveByPK($catalogItemId);
+		$connectedEvent = BaseScheduleEventPeer::retrieveByPK($this->scheduledEventId);
+
+		// validate that the catalogItem type is appropriate
+		if (!$vendorCatalogItem instanceof IVendorScheduledCatalogItem)
+		{
+			throw new KalturaAPIException(KalturaReachErrors::CATALOG_ITEM_AND_JOB_DATA_MISMATCH, get_class($vendorCatalogItem), 'KalturaScheduledVendorTaskData');
+		}
+
+		$startTime = $this->startDate ?: $connectedEvent->getStartDate(null);
```
Comment (inbal-ben-david) [inline plugins/reach/lib/api/data/entryVendorTaks/KalturaScheduledVendorTaskData.php:130]: can you explicitly wrote the value that returns on true

File: plugins/reach/lib/api/data/entryVendorTaks/KalturaScheduledVendorTaskData.php [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+		$minimalOrderTimeSec = $vendorCatalogItem->getMinimalOrderTime() * dateUtils::MINUTE;
+		if ($startTime - time() < $minimalOrderTimeSec)
+		{
+			throw new KalturaAPIException(KalturaReachErrors::TOO_LATE_ORDER, $this->scheduledEventId, $vendorCatalogItem->getId(), $vendorCatalogItem->getMinimalOrderTime());
+		}
+
+		$endTime = $this->endDate ?: $connectedEvent->getEndDate(null);
```
Comment (inbal-ben-david) [inline plugins/reach/lib/api/data/entryVendorTaks/KalturaScheduledVendorTaskData.php:137]: can you explicitly wrote the value that returns on true

File: plugins/reach/lib/kReachManager.php [hunk 7]
```diff
@@ -474,13 +495,32 @@ private function updateReachProfileCreditUsage(EntryVendorTask $entryVendorTask)
...
-		ReachProfilePeer::updateUsedCredit($entryVendorTask->getReachProfileId(), -$entryVendorTask->getPrice());
-		
-		//Rest task price so that reports will be aligned with the total used credit
-		$entryVendorTask->setOldPrice($entryVendorTask->getPrice());
-		$entryVendorTask->setPrice(0);
-		$entryVendorTask->save();
+		kReachUtils::refundTask($entryVendorTask);
+	}
+
+	private function handleAbortTask(EntryVendorTask $entryVendorTask)
```
Comment (inbal-ben-david) [inline plugins/reach/lib/kReachManager.php:507]: protected

File: plugins/reach/services/EntryVendorTaskService.php [hunk 8]
```diff
@@ -78,7 +78,7 @@ public function addEntryVendorTaskImpl($entryVendorTask, $taskVersion, $dbEntry,
 		$existingTask = EntryVendorTaskPeer::retrieveOneActiveOrCompleteTask($entryVendorTask->entryId, $entryVendorTask->catalogItemId, kCurrentContext::getCurrentPartnerId(), $taskVersion);
 		if ($existingTask)
 		{
-			if (!$dbVendorCatalogItem->getAllowResubmission() ||
+			if ((!$dbVendorCatalogItem->getAllowResubmission() && !$entryVendorTask->isScheduled()) ||
```
Comment (inbal-ben-david) [inline plugins/reach/services/EntryVendorTaskService.php:81]: this going to affect all tasks? not just live?

File: plugins/schedule/base/lib/model/LiveStreamScheduleEvent.php [hunk 9]
```diff
@@ -113,7 +115,7 @@ public function setLiveFeatures($v)
 	public function getLiveFeatures()
 	{
 		$serializedFeatures = $this->getFromCustomData(self::FEATURES_ARRAY);
-		return unserialize($serializedFeatures);
+		return unserialize($serializedFeatures) ?: array();
```
Comment (inbal-ben-david) [inline plugins/schedule/base/lib/model/LiveStreamScheduleEvent.php:118]: put explicit value for true

**Review by inbal-ben-david — COMMENTED** [review]
need to also add release notes

File: plugins/reach/services/EntryVendorTaskService.php (hunk 8 above)
Comment (nadav-ravid) [reply plugins/reach/services/EntryVendorTaskService.php:81]: No. only the live caption tasks currently return isScheduled true

File: plugins/reach/lib/kReachManager.php (hunk 7 above)
Comment (nadav-ravid) [reply plugins/reach/lib/kReachManager.php:507]: That would be incosistent with the pre existing handleErrorTask. do you want me to change it anyway?

File: plugins/reach/lib/model/enums/EntryVendorTaskStatus.php [hunk 10]
```diff
@@ -13,4 +13,5 @@ interface EntryVendorTaskStatus extends BaseEnum
 	const ERROR 				= 6;
 	const ABORTED 				= 7;
 	const PENDING_ENTRY_READY	= 8;
+	const SCHEDULED 			= 9;
```
[the remark this answers is not in the document]
Comment (nadav-ravid) [reply plugins/reach/lib/model/enums/EntryVendorTaskStatus.php:16]: line above is too long to align properly without making the actual code look worse

File: plugins/reach/lib/kReachManager.php (hunk 7 above)
Comment (inbal-ben-david) [reply plugins/reach/lib/kReachManager.php]: no, if thats the case keep it

**Comment by nadav-ravid** [comment]
we are intentionally.not adding the script to the release notes because we want to run the script manually once we release the feature.

**Pull request merged**